File: Solutions/395_Longest_Substringwith_At_Least_K_Repeating_Characters/395_Longest_Substringwith_At_Least_K_Repeating_Characters.go
//https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/discuss/308459/golang
func longestSubstring(s string, k int) int {
	if k <= 0 {
		return 0
	}
	cache := make(map[string]int)
	return longestSubstringHelper(s, k, cache)
}

func longestSubstringHelper(s string, k int, cache map[string]int) int {
	if r, ok := cache[s]; ok {
		return r
	}
	res := 0
	counts := make([]int, 26)
	for i := 0; i < len(s); i++ {
		counts[int(s[i]-'a')]++
	}

	meet := true
	for i := 0; i < 26; i++ {
		if counts[i] > 0 && counts[i] < k {
			meet = false
			strs := strings.Split(s, string(i+'a'))
			for j := 0; j < len(strs); j++ {
				res = max(res, longestSubstringHelper(strs[j], k, cache))
			}
		}
	}

	if meet {
		cache[s] = len(s)
		return len(s)
	}
	cache[s] = res
	return res
}
func max(x, y int) int {
	if x > y {
		return x
	}

	return y
}

//Fastest
func longestSubstring(s string, k int) int {
    var m [26]int
    num := 0
    for i:=0; i<len(s); i++ {
        idx := int(s[i]-'a')
        if m[idx] == 0 {
            num++
        }
        m[idx]++
    }
    
    max := 0
    
    for n:=1; n<=num; n++ {
        var m [26]int
        cnt := 0
        i := 0
        j := 0
        unique := 0
        
        for j < len(s) {
            
            if unique <= n {
                c := int(s[j]-'a')
                m[c]++
                j++
                // cnt is the number that >= k
                if m[c] == k {
                    cnt++
                }
                if m[c] == 1 {
                    unique++
                }
            } else {
                c := int(s[i]-'a')
                if m[c] == k {
                    cnt--
                }
                m[c]--
                i++
                if m[c] == 0 {
                    unique--
                }
            }
            
            if unique == n && unique == cnt {
                if j-i > max {
                    max = j-i
                }
            }
        }
    }
    
    return max   
}

File: Solutions/395_Longest_Substringwith_At_Least_K_Repeating_Characters/395_Longest_Substringwith_At_Least_K_Repeating_Characters.js
//Third fastest
//Split by regex javascript solution -> https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/discuss/408410/Split-by-regex-javascript-solution-56-ms
var longestSubstring = function(s, k) {
    var max = 0;
    
    var find = function (str) {
        var map = new Map();
        for (var i = 0; i < str.length; i++) {
            map.set(str[i], (map.get(str[i]) || 0) + 1);
        }
        
        var split = [];
        for (var [key, count] of map) {
            if (count < k) {
                split.push(key);
            }
        }
        
        if (split.length === 0) {
            if (str.length > max) {
                max = str.length;
            }
            return;
        }
        
        var substr = str.split(new RegExp("[" + split.join("") + "]+"))
        for (var i = 0; i < substr.length; i++) {
            if (substr[i].length >= k && substr[i].length > max) {
                find(substr[i]);
            }
        }
    };
    
    find(s);
    
    return max;
};

//Second fastest
var longestSubstring = function(s, k) {
    let result = 0;
    let maxUnique = 0;
    const countMap = new Array(26).fill(0);
    // count number of unique letters
    for (let i=0; i<s.length; i++) {
        if (countMap[s.charCodeAt(i)-97] == 0) {
            countMap[s.charCodeAt(i)-97]++;
            maxUnique++;
        }
    }
    for (let currentUnique=1; currentUnique<=maxUnique; currentUnique++) {
        // reset count map
        countMap.fill(0);
        let windowStart=0;
        let windowEnd=0;
        let index=0;
        let unique=0;
        let atLeastK=0;
        while (windowEnd < s.length) {
            // expand sliding window
            if (unique <= currentUnique) {
                index=s.charCodeAt(windowEnd)-97;
                if (countMap[index]==0) {
                    unique++;
                }
                countMap[index]++;
                if (countMap[index]==k) {
                    atLeastK++;
                }
                windowEnd++;
            } else {
                // shrink sliding window
                index=s.charCodeAt(windowStart) - 97;
                if (countMap[index]==k) {
                    atLeastK--;
                }
                countMap[index]--;
                if (countMap[index]==0) {
                    unique--;
                }
                windowStart++;
            }
            if (unique == currentUnique && unique == atLeastK) {
                result = Math.max(windowEnd-windowStart, result);
            }
        }
    }
    return result;
};

//First fastest
// https://leetcode.com/problems/longest-substring-with-at-least-k-repeating-characters/discuss/87739/Java-Strict-O(N)-Two-Pointer-Solution

const CODE_A = 'a'.charCodeAt(0);

function longestSubstringWithTargetUniqueNum(s, k, targetUniqueNum) {
    const map = new Array(26).fill(0);
    let unique = 0;
    let noLessThanK = 0;
    let begin = 0, end = 0;

    let ret = 0;

    while (end < s.length) {
        const idxEnd = s[end].charCodeAt(0) - CODE_A;
        if (map[idxEnd] === 0) {
            unique++;
        }
        map[idxEnd]++;

        if (map[idxEnd] === k) {
            noLessThanK++;
        }
        end++;

        while (unique > targetUniqueNum) {
            const idxBegin = s[begin].charCodeAt(0) - CODE_A;
            if (map[idxBegin] === k) {
                noLessThanK--;
            }

            map[idxBegin]--;

            if (map[idxBegin] === 0) {
                unique--;
            }

            begin++;
        }

        if (unique === targetUniqueNum && noLessThanK === unique) {
            // here, end has been incremented by "end++". So current "end" is next "end", and we don't need to add 1
            const diff = end - begin;
            if (diff > ret) {
                ret = diff;
            }
        }
    }

    return ret;
}

/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var longestSubstring = function (s, k) {
    let ret = 0;
    for (let i = 1; i <= 26; i++) {
        const curRet = longestSubstringWithTargetUniqueNum(s, k, i);
        if (curRet > ret) {
            ret = curRet;
        }
    }
    // console.log(ret);
    return ret;
};

// longestSubstring('aaabb', 3); // 3
// longestSubstring('ababbc', 2); // 5